src/neuroglancer/datasource/graphene/frontend.ts

```diff
@@ -0,0 +1,746 @@
+  getFragmentKey(objectKey: string|null, fragmentId: string) {
+    objectKey;
+    return getGrapheneFragmentKey(fragmentId);
+  }
+}
+
+function resolvePath(a: string, b: string) {
+  const outputParts = a.split('/');
+  for (const part of b.split('/')) {
+    if (part === '..') {
+      if (outputParts.length !== 0) {
+        outputParts.length = outputParts.length - 1;
+        continue;
+      }
+    }
+    outputParts.push(part);
+  }
+  return outputParts.join('/');
+}
+
+class AppInfo {
+  segmentationUrl: string;
+  meshingUrl: string;
+  supported_api_versions: number[];
+  constructor(infoUrl: string, obj: any) {
+    // .../1.0/... is the legacy link style
+    // .../table/... is the current, version agnostic link style (for retrieving the info file)
+    const linkStyle = /^(https?:\/\/[.\w:\-\/]+)\/segmentation\/(?:1\.0|table)\/([^\/]+)\/?$/;
+    let match = infoUrl.match(linkStyle);
+    if (match === null) {
+      throw Error(`Graph URL invalid: ${infoUrl}`);
+    }
+    this.segmentationUrl = `${match[1]}/segmentation/api/v${PYCG_APP_VERSION}/table/${match[2]}`;
+    this.meshingUrl = `${match[1]}/meshing/api/v${PYCG_APP_VERSION}/table/${match[2]}`;
+
+    try {
+      verifyObject(obj);
+      this.supported_api_versions = verifyObjectProperty(
+          obj, 'supported_api_versions', x => parseArray(x, verifyNonnegativeInt));
+    } catch (error) {
+      // Dealing with a prehistoric graph server with no version information
+      this.supported_api_versions = [0];
+    }
+    if (PYCG_APP_VERSION in this.supported_api_versions === false) {
+      const redirectMsgBox = new StatusMessage();
+      const redirectMsg = `This Neuroglancer branch requires Graph Server version ${
+          PYCG_APP_VERSION}, but the server only supports version(s) ${
+          this.supported_api_versions}.`;
+
+      if (location.hostname.includes('neuromancer-seung-import.appspot.com')) {
+        const redirectLoc = new URL(location.href);
+        redirectLoc.hostname = `graphene-v${
+            this.supported_api_versions.slice(-1)[0]}-dot-neuromancer-seung-import.appspot.com`;
+        redirectMsgBox.setHTML(`Try <a href="${redirectLoc.href}">${redirectLoc.hostname}</a>?`);
+      }
+      throw new Error(redirectMsg);
+    }
+  }
+}
+
+const N_BITS_FOR_LAYER_ID_DEFAULT = 8;
+
+class GraphInfo {
+  chunkSize: vec3;
+  nBitsForLayerId: number;
+  constructor(obj: any) {
+    verifyObject(obj);
+    this.chunkSize = verifyObjectProperty(
+        obj, 'chunk_size', x => parseFixedLengthArray(vec3.create(), x, verifyPositiveInt));
+    this.nBitsForLayerId = verifyOptionalObjectProperty(
+        obj, 'n_bits_for_layer_id', verifyPositiveInt, N_BITS_FOR_LAYER_ID_DEFAULT);
+  }
+}
+
+interface GrapheneMultiscaleVolumeInfo extends MultiscaleVolumeInfo {
+  dataUrl: string;
+  app: AppInfo;
+  graph: GraphInfo;
+}
+
+function parseSpecialUrlOld(url: string): string { // TODO: brought back old parseSpecialUrl
```

Comment:
What is the reason for this?

Reply:
It looks like it was needed because we hadn't adopted cancellableFetchSpecialOk which does the transformation from gs to https. It seems to be working correctly if I swap to that function and use undefined for the credentialsProvider but this goes back to your original comment about a function where I had previously set the credentialsProvider to undefined if the url was GCS. I need to confirm when we may need credentials for such requests.